cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java

```diff
@@ -107,28 +117,53 @@ public RuntimeContext call() throws Exception {
   /**
    * Adds a listener to the {@link ProgramController} and blocks for completion.
    *
+   *
+   * @param programId the {@link Id.Program} of the program to be run
    * @param controller the {@link ProgramController} for the program
    * @param context    the {@link RuntimeContext}
    * @return {@link RuntimeContext} of the completed program
    * @throws Exception if the execution failed
    */
-  protected RuntimeContext executeProgram(final ProgramController controller,
+  protected RuntimeContext executeProgram(final Id.Program programId, final ProgramController controller,
                                           final RuntimeContext context) throws Exception {
+    // TODO: Perhaps generate a different runId for each program contained in a workflow
+    final String runId = controller.getRunId().getId();
     // Execute the program.
     final SettableFuture<RuntimeContext> completion = SettableFuture.create();
     controller.addListener(new AbstractListener() {
       @Override
+      public void init(ProgramController.State state, @Nullable Throwable cause) {
+        store.setStart(programId, runId, TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS));
```

Comment:
Why?? The existence of Controller and listener is exactly to avoid logic like this inside ProgramRunner. The ProgramRunner should only cares about the state transition, but shouldn't responsible for changing things in Store.